internal/users/manager.go

```diff
@@ -292,33 +383,254 @@ func (m *Manager) AllShadows() ([]ShadowEntry, error) {
-	return generateID(str, m.config.UIDMin, m.config.UIDMax)
+// RegisterUserPreAuth registers a temporary user with the given name and a unique UID in our NSS handler (in memory,
+// not in the database). It returns the generated UID or an error if the user could not be registered.
+func (m *Manager) RegisterUserPreAuth(name string) (uint32, error) {
+	m.temporaryEntriesMu.Lock()
+	defer m.temporaryEntriesMu.Unlock()
+
+	if m.numPreAuthUsers >= maxPreAuthUsers {
+		return 0, fmt.Errorf("maximum number of pre-auth users reached, login for new users via SSH is disabled, restart authd to enable it")
+	}
+
+	uid, _, err := m.registerUser(name, true)
+	if err != nil {
+		return 0, fmt.Errorf("could not register user %q: %w", name, err)
+	}
+
+	m.numPreAuthUsers++
+
+	return uid, nil
 }
 
-// GenerateGID deterministically generates an ID between from the given string, ignoring case,
-// in the range [GIDMin, GIDMax]. The generated ID is *not* guaranteed to be unique.
-func (m *Manager) GenerateGID(str string) uint32 {
-	return generateID(str, m.config.GIDMin, m.config.GIDMax)
+// registerUser registers a temporary user with the given name and a unique UID in our NSS handler (in memory, not in
+// the database).
+//
+// The caller must lock m.temporaryEntriesMu for writing before calling this function.
+//
+// Returns the generated UID and a cleanup function that should be called to remove the temporary user once the user was
+// added to the database.
+func (m *Manager) registerUser(name string, preAuth bool) (uid uint32, cleanup func(), err error) {
+	// Check if there is already a temporary user for that name
+	for uid, user := range m.temporaryUsers {
+		if user.name == name+"-preauth" {
+			// A temporary user with the same name is already registered. To avoid that we generate multiple UIDs for
+			// the same user, we return the already generated UID.
+			cleanup = func() { m.deleteTemporaryUser(uid) }
+			return uid, cleanup, nil
+		}
+	}
+
+	for {
+		uid, err = m.GenerateUID()
+		if err != nil {
+			return 0, nil, err
+		}
+
+		// To avoid races where a user with this UID is created by some NSS source after we checked, we register this
+		// UID in our NSS handler and then check if another user with the same UID exists in the system. This way we
+		// can guarantee that the UID is unique, under the assumption that other NSS sources don't add users with a UID
+		// that we already registered (if they do, there's nothing we can do about it).
+		var tmpName string
+		tmpName, cleanup, err = m.addTemporaryUser(uid, preAuth)
+		if errors.Is(err, errUserAlreadyExists) {
+			log.Debugf(context.Background(), "UID %d already in use, generating a new one", uid)
+			continue
+		}
+		if err != nil {
+			return 0, nil, fmt.Errorf("could not register temporary user: %w", err)
+		}
+
+		if unique := m.checkIsUniqueUID(uid, tmpName); unique {
+			if preAuth {
+				// Rename the temporary user so that we can recognize it later.
+				tmpName = name + "-preauth"
```

Comment:
This looks like a bit fragile to me... :thinking:

Reply:
I reworked that part